Add compiler testrunner and unsafe keywords to LCB

runrevmark opened the pull request:
This patch adds a new test runner for running LCB compiler tests.

The compiler tests are run by using 'make compiler-check' in the
tests folder.

The compiler test runner recursively searches for tests in the
tests folder, processing any file with a 'compilertest' extension.

Compiler tests are text files containing a number of test clauses.
Each test clause is named, and consists of a fragment of LCB code
to be parsed to the compiler, along with an expectation and list
of errors and/or warnings to check for.

The compiler test format uses lines beginning with %% to indicate
comments.

A test clause has the following structure:

```
%TEST <name>
... LCB code ...
%EXPECT ( PASS | FAIL | XPASS | SKIP )
... assertions ...
%ENDTEST
```

Here the assertion clauses can be one of the following:

```
%SUCCESS
%ERROR <error-id> AT <pos-id>
%WARNING <warning-id> AT <pos-id>
```

In the body of LCB code for each test clause, %{pos-id} can be used
to indicate a named position - these can be used in error and warning
assertions.

When running a test clause, the test runner will first extract the
LCB code, removing any named positions from it. It will then attempt
to compile the code. The output of the compiler is checked against
the list of expected errors and warnings, or success and each test
will succeed or fail as a result, in line with the expectation.

Additionally, there is a new compiler option '--error-info' which
emits error information in a more structured way. Each error or
warning that is emitted has the following format:

```
<file>:<line>:<col>:warning|error:<id>:<param>
```

Here, rather than use a human readable message, the compiler emits
the error's id, and optional parameter separately.

---

This patch adds the idea of unsafe handlers and unsafe blocks
to LCB.

runrevmark committed:
[[ UnsafeAttrib ]] Add unsafe handlers and unsafe blocks

This patch adds the idea of unsafe handlers and unsafe blocks
to LCB.

peter-b commented on tests/Makefile:
```diff
@@ -26,7 +26,8 @@ guess_platform_script := \
 
-bin_dir ?= $(top_srcdir)/$(guess_platform)-bin
+#bin_dir ?= $(top_srcdir)/$(guess_platform)-bin
+bin_dir ?= $(top_srcdir)/_build/mac/Debug
```
You'll need to figure out what to do with this, or it'll break running the tests on non-Mac.

peter-b commented on tests/Makefile:
```diff
@@ -141,6 +148,17 @@ lcs-check: $(LCS_ENGINE)
+	@rm -f $(LCS_LOG)
+	@cmd="$(LCC_CMD)"; \
+	echo "$$cmd" $(_PRINT_RULE); \
+	$$cmd
```
- I'm not sure that this rule needs to have `($LCS_ENGINE)` as a dependency
- I would normally incorporate setting the environment variable into the script
- I think you're deleting the LiveCode Script tests' logfile -- maybe define an `LCC_LOG` make variable

``` make
compiler-check:
    @rm -f $(LCC_LOG)
    @export LC_COMPILE="$(LC_COMPILE)"; \
    cmd="$(LCC_CMD)"; \
    echo "$$cmd" $(_PRINT_RULE); \
    $$cmd
```

peter-b commented on tests/_compilertestrunner.livecodescript:
```diff
@@ -0,0 +1,637 @@
+﻿script "CompilerTestRunner"
+/*
+Copyright (C) 2015 LiveCode Ltd.
```
2016

peter-b commented on tests/_compilertestrunner.livecodescript:
```diff
@@ -0,0 +1,637 @@
+along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
+
+-- FIXME provide this on the command line
+constant kLogFilename = "_test_suite.log"
```
Probably need to call this "_compiler_test_suite.log" to avoid a conflict with the LCS tests.  Or fix both of them to accept a `--logfile` argument.

peter-b commented on tests/_compilertestrunner.livecodescript:
```diff
@@ -0,0 +1,637 @@
+on startup
+   try
+      if $LC_COMPILE is empty then
+         throw "LC_COMPILE environment var must be set to lc-compile to use"
```
I think it makes sense to throw this error somewhere in `CompilerTestRunnerMain` at the point where you know whether or not you need `$LC_COMPILE` to be set for this invocation (for example, using the `help` command doesn't require `$LC_COMPILE` to be set in the environment).

peter-b replied:
You could also consider passing the `lc-compile` on the command line.

peter-b commented on tests/_testerlib.livecodescript:
```diff
@@ -138,6 +188,14 @@ function TesterGetPrettyTestName pFilename
+      set the itemDelimiter to "."
+      return item 1 to -2 of pFileName
+   end if
+end TesterGetPrettyCompilerTestName
```
Rather than adding this function, I suggest that you amend `TesterGetPrettyTestName()` so that it takes the suffix to remove as an argument.

peter-b commented on tests/_testerlib.livecodescript:
```diff
@@ -41,13 +41,30 @@ function TesterGetTestFileNames pBaseFolder
+   TesterGetActualFileNames_Recursive pBaseFolder, empty, "compilertest", tFiles, tCount
+
+   return tFiles
+end TesterGetCompilerTestFileNames
```
I suggest adding an optional `pExtension` argument to `TesterGetTestFileNames()` instead of adding this function (which is highly duplicative).

peter-b commented on tests/_compilertestrunner.livecodescript:
```diff
@@ -0,0 +1,637 @@
+
+----------------------------------------------------------------
+-- Logging helpers
+----------------------------------------------------------------
```
I suggest renaming the logging helpers to have a `Tester` prefix and moving them into `_testerlib.livecodescript`.

runrevmark replied:
These are mostly specific to the compilerrunner at the moment - I've made it so that you can set the LCC_VERBOSE env var to 1 to emit all the info about what it finds in the compiler tests.

peter-b commented on tests/_testerlib.livecodescript:
```diff
@@ -130,6 +147,39 @@ function TesterParseTestCommandNames pFilename
+   end repeat
+
+   return tTestNames
+end TesterParseCompilerTestNames
```
This seems quite compiler-test-specific, so why not keep it in the compiler test runner for now?  The `_testerlib.livecodescript` library was added to contain code that's shared between the IDE & engine test runner implementations, but this function isn't likely to be useful elsewhere.

peter-b commented on toolchain/lc-compile/src/main.c:
```diff
@@ -102,20 +103,22 @@ usage(int status)
+"      --manifest MANIFEST    Filename for generated manifest.\n"
+"      --interface INTERFACE  Filename for generated interface.\n"
+"      -Werror                Turn all warnings into errors.\n"
+"      --error-info           Emit warnings and errors in form suitable for testing.\n"
```
I would still really prefer making the "normal" warning & error format more suitable for automated processing. 😛

peter-b commented on toolchain/lc-compile/src/report.c:
```diff
@@ -26,6 +26,8 @@ extern int IsDependencyCompile(void);
 
 ////////////////////////////////////////////////////////////////////////////////
 
+int ErrorInfo = 0;
```
Ideally, this should be called `s_is_<something>` for consistency.

peter-b replied:
Also, it should be initialised in `InitializeReports`.  BTW I just noticed that `s_verbose_level` is statically initialised to `1` and then reset to `0` during report initialization -- I think that's an artefact of the bytecode or lexical scope patches.  Probably the globals should either be statically initialised, or initialised by `InitializeReports`, but not both.

runrevmark replied:
Most of the 'option' globals in lc-compile are of the form OutputAsC / ErrorInfo - so there is some inconsistency here. 's__' is usually used a compilation-unit local var, so those which are shared amongst units should really be 'g__'. I'll leave this as ErrorInfo for now, and tweak the initialization. We can clean up this naming in a different patch uniformly.

peter-b commented on tests/_compilertestrunner.livecodescript:
```diff
@@ -0,0 +1,681 @@
+      if doesCompilerOutputSatisfyAssertion(tCompilerOutput, tCompilerExitStatus, tAssertion, tTestInfo["positions"]) then
+         put "ok # " after tTestOutput
+      else
+         put "not ok # " after tTestOutput
```
I'm not sure this matches the way we generate TAP output elsewhere in the testsuite.

`compilerPrettyPrintAssertion()` appears to return a message describing the test. I think that elsewhere, we use:

```
ok - <description>
not ok - <description>
```

That seems to match my understanding of the TAP specification, too.

runrevmark committed:
[[ CompilerTests ]] Add a compiler test runner

This patch adds a new test runner for running LCB compiler tests.

The compiler tests are run by using 'make compiler-check' in the
tests folder.

The compiler test runner recursively searches for tests in the
tests folder, processing any file with a 'compilertest' extension.

Compiler tests are text files containing a number of test clauses.
Each test clause is named, and consists of a fragment of LCB code
to be parsed to the compiler, along with an expectation and list
of errors and/or warnings to check for.

The compiler test format uses lines beginning with %% to indicate
comments.

A test clause has the following structure:

    %TEST <name>
    ... LCB code ...
    %EXPECT ( PASS | FAIL | XPASS | SKIP )
    ... assertions ...
    %ENDTEST

Here the assertion clauses can be one of the following:

    %SUCCESS
    %ERROR <error-id> AT <pos-id> [ WITH <hint> ]
    %WARNING <warning-id> AT <pos-id> [ WITH <hint> ]

In the body of LCB code for each test clause, %{pos-id} can be used
to indicate a named position - these can be used in error and warning
assertions.

When running a test clause, the test runner will first extract the
LCB code, removing any named positions from it. It will then attempt
to compile the code. The output of the compiler is checked against
the list of expected errors and warnings, or success and each test
will succeed or fail as a result, in line with the expectation.

Additionally, there is a new compiler option '--error-info' which
emits error information in a more structured way. Each error or
warning that is emitted has the following format:

    <file>:<line>:<col>:warning|error:<id>:<hint>

Here, rather than use a human readable message, the compiler emits
the error's id, and optional parameter separately.
